graphula-core/src/Graphula.hs

```diff
@@ -429,85 +467,234 @@ class HasDependencies a where
         (toEot a)
         (toEot dependencies)
 
-data GenerationFailure =
-  GenerationFailureMaxAttempts TypeRep
+data KeySourceType
+  = SourceDefault
+  -- ^ Generate keys using the database's @DEFAULT@ strategy
+  | SourceArbitrary
+  -- ^ Generate keys using the @'Arbitrary'@ instance for the @'Key'@
+  | SourceExternal
+  -- ^ Always explicitly pass an external key
+
+-- | Handle key generation for @'SourceDefault'@ and @'SourceArbitrary'@
+--
+-- Ths could be a single-parameter class, but carrying the @a@ around
+-- lets us give a better error message when @'node'@ is called instead
+-- of @'nodeKeyed'@.
+--
+class GenEntityKey (s :: KeySourceType) a where
+  type KeyConstraint s a :: Constraint
```

Comment:
This is pretty interesting. I don't think I've seen `Constraint` associated type families :eyes:

Reply:
Lol, I think you've forgotten that _you_ added

```haskell
class MonaGraphulaBackend m where
  type Logging m :: * -> Constraint
  type Generate m :: * -> Constraint
```

Reply:
So I did!

`i-have-no-memory-of-this-place.gif`